credentials/gcp.go

```diff
@@ -0,0 +1,187 @@
+// in the OS keyring
+const KeyNamespace = "scuttle"
+
+// CredentialVar is a reference to the google application credential env configuration path
+// which can be used to short-circuit the credential provider storage and read directly from
+// the provided credential.json
+const CredentialVar = "GOOGLE_APPLICATION_CREDENTIALS"
+
+// GoogleCredential represents the workflow needed to attain a google client credential
+// from the google API. The resulting credential will be a RefreshToken scoped to the
+// google cloud platform KMS api.
+type GoogleCredential struct {
+	Name string
+}
+
+// GoogleToken is the data structure to be used when serializing to token storage.
+// This json object contains all the details needed to configure the client.
+type GoogleToken struct {
+	ClientID     string `json:"client_id"`
+	ClientSecret string `json:"client_secret"`
+	RefreshToken string `json:"refresh_token"`
+	TheType      string `json:"type"`
+}
+
+// GoogleClientJSON is the wrapping object in a downloaded client configuration.
+type GoogleClientJSON struct {
+	ClientConfig GoogleClientConfig `json:"installed"`
+}
+
+// GoogleClientConfig is a struct wrapping the credential detail needed to provide sctl to identify
+// itself to the google API.
+type GoogleClientConfig struct {
+	AuthProviderX509CertURL string   `json:"auth_provider_x509_cert_url"`
+	AuthURI                 string   `json:"auth_uri"`
+	ClientID                string   `json:"client_id"`
+	ClientSecret            string   `json:"client_secret"`
+	ProjectID               string   `json:"project_id"`
+	RedirectUris            []string `json:"redirect_uris"`
+	TokenURI                string   `json:"token_uri"`
+}
+
+// DeleteCredential removes a stored credential from the system keystore, and
+// will remove the assigned defualt credential from configuration.
+func (gc GoogleCredential) DeleteCredential(credentialName string) error {
+	return keyring.Delete(KeyNamespace, credentialName+"-gcp")
+}
+
+// SaveCredential marshals the received GoogleToken and stores the resulting json blob in the
+// system keystore for secure storage at rest.
+func (gc GoogleCredential) SaveCredential(credentialName string, credential GoogleToken) error {
+	jsonData, err := json.Marshal(credential)
+	if err != nil {
+		return err
+	}
+	return keyring.Set(KeyNamespace, credentialName+"-gcp", string(jsonData))
+}
+
+// GetCredential returns a decoded GoogleToken from the system keystore. The resulting
+// object is serializeable and should be used in conjunction with API Options configFromJSON()
+func (gc GoogleCredential) GetCredential(credentialName string) (GoogleToken, error) {
+	userTokenJSON, err := keyring.Get(KeyNamespace, credentialName+"-gcp")
+	if err != nil {
+		return GoogleToken{}, err
+	}
+	var token GoogleToken
+	err = json.Unmarshal([]byte(userTokenJSON), &token)
+	if err != nil {
+		return GoogleToken{}, err
+	}
+
+	return token, nil
```

Comment:
I'm wondering if these functions belong on the `GoogleCredential` struct or not. Maybe it would be good to keep them there organizationally. My thought stems from the fact that they don't do anything with the `GoogleCredential` state, and to call any of these methods, you'd basically need to initialize an empty struct just to call them, e.g.

```go
var cred credentials.GoogleCredential
return cred.DeleteCredential("default")
```

when doing something like 
```go
return DeleteCredential("default")
```
should suffice? I guess what I'm saying is that these methods feel like they are more attached to the keyring than to the google credential.